File: internal/observability/topic_traffic.go
package observability

import (
	"context"
	"sync"
	"sync/atomic"

	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/metric"
)

// TopicTrafficMeter holds the always-emit per-topic Produce/Fetch
// counters that drive Grafana's "records in / records out per
// topic" panels. gh #115 + gh #121 PR1.
//
// Apache Kafka exposes BytesInPerSec / BytesOutPerSec as MBean
// meters keyed by topic — they have a "current value" at every
// scrape, even when the topic is idle, so dashboard panels show a
// flat zero line instead of a gap. Skafka pre-#121 used OTel
// Int64Counter instruments that only emit when Add() fires; idle
// topics disappeared from the timeseries entirely.
//
// Design:
//
//  1. Per-topic atomic accumulators, set up at topic-discovery
//     time via Touch(). Hot-path Produce/Fetch handlers update
//     them with one atomic.Add — fewer allocations than the old
//     `metric.WithAttributes(...)` per call.
//  2. The OTel SDK registers 4 ObservableInt64Counter instruments
//     and calls a single callback at every scrape interval. The
//     callback walks the accumulator map and emits one cumulative
//     observation per topic per metric. Observable counters always
//     emit, even when the underlying value hasn't changed since
//     the last scrape — that's the gh #115 fix.
//
// Topics that have NEVER received traffic but exist in the
// TopicRegistry still emit zero — callers should invoke Touch()
// for every topic the broker becomes aware of (TopicRegistry.Add
// hook). This guarantees a dashboard panel shows every existing
// topic, traffic or not.
type TopicTrafficMeter struct {
	mu       sync.RWMutex
	counters map[string]*topicCounters
}

type topicCounters struct {
	produceRecords atomic.Int64
	produceBytes   atomic.Int64
	fetchRecords   atomic.Int64
	fetchBytes     atomic.Int64
}

// NewTopicTrafficMeter constructs the meter. The returned meter is
// usable immediately; the OTel instruments are wired separately by
// registerTopicTrafficInstruments.
func NewTopicTrafficMeter() *TopicTrafficMeter {
	return &TopicTrafficMeter{counters: make(map[string]*topicCounters)}
}

// Touch ensures an accumulator entry exists for the topic. Idempotent.
// Called when a topic becomes known to the broker (TopicRegistry.Add).
// After this, the topic emits zero on every scrape until traffic
// arrives — eliminating the gh #115 "no data" gap on idle topics.
func (m *TopicTrafficMeter) Touch(topic string) {
	if topic == "" {
		return
	}
	m.mu.RLock()
	_, ok := m.counters[topic]
	m.mu.RUnlock()
	if ok {
		return
	}
	m.mu.Lock()
	if _, ok := m.counters[topic]; !ok {
		m.counters[topic] = &topicCounters{}
	}
	m.mu.Unlock()
}

// Forget removes the accumulator. Called when a topic is deleted so
// orphan timeseries don't linger on the dashboard indefinitely.
// Idempotent.
func (m *TopicTrafficMeter) Forget(topic string) {
	m.mu.Lock()
	delete(m.counters, topic)
	m.mu.Unlock()
}

// RecordProduce is the hot-path call from the Produce handler.
// Auto-touches the topic — callers don't have to invoke Touch
// separately for first-Produce-on-a-new-topic.
func (m *TopicTrafficMeter) RecordProduce(topic string, records, bytes int64) {
	tc := m.ensure(topic)
	if tc == nil {
		return
	}
	if records > 0 {
		tc.produceRecords.Add(records)
	}
	if bytes > 0 {
		tc.produceBytes.Add(bytes)
	}
}

// RecordFetch is the hot-path call from the Fetch handler.
// Auto-touches the topic. bytes may be 0 (empty Fetch response);
// the call still updates the timestamp on the underlying timeseries
// so dashboards see "fetch happened, just empty".
func (m *TopicTrafficMeter) RecordFetch(topic string, records, bytes int64) {
	tc := m.ensure(topic)
	if tc == nil {
		return
	}
	if records > 0 {
		tc.fetchRecords.Add(records)
	}
	if bytes > 0 {
		tc.fetchBytes.Add(bytes)
	}
}

// ensure returns the accumulator for `topic`, creating it if missing.
// Fast-path is a read-lock; only the first-time-seen case takes the
// write lock.
func (m *TopicTrafficMeter) ensure(topic string) *topicCounters {
	if topic == "" {
		return nil
	}
	m.mu.RLock()
	tc, ok := m.counters[topic]
	m.mu.RUnlock()
	if ok {
		return tc
	}
	m.mu.Lock()
	defer m.mu.Unlock()
	if tc, ok := m.counters[topic]; ok {
		return tc
	}
	tc = &topicCounters{}
	m.counters[topic] = tc
	return tc
}

// snapshot returns a slice of (topic, counters) pairs for the
// callback. Holds the read lock just long enough to copy the slice
// header — no per-topic copying needed because the accumulator
// pointers are stable across map shape changes.
func (m *TopicTrafficMeter) snapshot() []topicSnapshot {
	m.mu.RLock()
	defer m.mu.RUnlock()
	out := make([]topicSnapshot, 0, len(m.counters))
	for name, tc := range m.counters {
		out = append(out, topicSnapshot{
			topic:          name,
			produceRecords: tc.produceRecords.Load(),
			produceBytes:   tc.produceBytes.Load(),
			fetchRecords:   tc.fetchRecords.Load(),
			fetchBytes:     tc.fetchBytes.Load(),
		})
	}
	return out
}

type topicSnapshot struct {
	topic          string
	produceRecords int64
	produceBytes   int64
	fetchRecords   int64
	fetchBytes     int64
}

// registerTopicTrafficInstruments wires the ObservableInt64Counters
// on the given meter and registers a single callback that walks the
// accumulator map and emits one observation per topic per metric.
// Returns the four instruments so callers can hold them if needed
// (currently only the callback uses them).
//
// Idempotent only across distinct meters — calling twice on the
// same meter panics (OTel forbids duplicate instrument names).
func registerTopicTrafficInstruments(m metric.Meter, meter *TopicTrafficMeter) error {
	produceRecords, err := m.Int64ObservableCounter("skafka.produce.records",
		metric.WithDescription("Cumulative records produced per topic. Emits at every scrape interval (including 0 for idle topics) so dashboard rate() panels never gap."),
		metric.WithUnit("{record}"))
	if err != nil {
		return err
	}
	produceBytes, err := m.Int64ObservableCounter("skafka.produce.bytes",
		metric.WithDescription("Cumulative bytes produced per topic. Idle-emit invariant."),
		metric.WithUnit("By"))
	if err != nil {
		return err
	}
	fetchRecords, err := m.Int64ObservableCounter("skafka.fetch.records",
		metric.WithDescription("Cumulative records fetched per topic. Idle-emit invariant."),
		metric.WithUnit("{record}"))
	if err != nil {
		return err
	}
	fetchBytes, err := m.Int64ObservableCounter("skafka.fetch.bytes",
		metric.WithDescription("Cumulative bytes fetched per topic. Idle-emit invariant."),
		metric.WithUnit("By"))
	if err != nil {
		return err
	}
	_, err = m.RegisterCallback(func(_ context.Context, obs metric.Observer) error {
		for _, snap := range meter.snapshot() {
			attrs := metric.WithAttributes(attribute.String("topic", snap.topic))
			obs.ObserveInt64(produceRecords, snap.produceRecords, attrs)
			obs.ObserveInt64(produceBytes, snap.produceBytes, attrs)
			obs.ObserveInt64(fetchRecords, snap.fetchRecords, attrs)
			obs.ObserveInt64(fetchBytes, snap.fetchBytes, attrs)
		}
		return nil
	}, produceRecords, produceBytes, fetchRecords, fetchBytes)
	return err
}

File: crates/kaas-observability/src/topic_traffic.rs
//! Per-topic Produce / Fetch counters that always emit at scrape.
//!
//! Per-topic traffic counters (gh #115, gh #121 PR1). Apache Kafka's `BytesInPerSec` / `BytesOutPerSec`
//! MBeans emit a current cumulative value at every scrape (idle topics
//! read zero, not "no data"). Pre-#121 kaas used `Int64Counter`
//! instruments that only emit on `Add()`, so idle topics disappeared
//! from the timeseries and `rate()` panels gapped.
//!
//! Design: per-topic atomic accumulators, updated by one atomic-add on
//! the hot path. The 4 [`ObservableCounter`] instruments each carry
//! their own callback that walks a snapshot of the map and observes
//! every topic at every scrape — even topics that haven't seen traffic
//! since boot.

use std::sync::atomic::{AtomicI64, Ordering};
use std::sync::Arc;

use dashmap::DashMap;
use opentelemetry::metrics::Meter;
use opentelemetry::KeyValue;

/// Always-emit per-topic accumulators + the OTel instrument bindings.
#[derive(Debug, Default)]
pub struct TopicTrafficMeter {
    counters: DashMap<String, Arc<TopicCounters>>,
}

#[derive(Debug, Default)]
struct TopicCounters {
    produce_records: AtomicI64,
    produce_bytes: AtomicI64,
    fetch_records: AtomicI64,
    fetch_bytes: AtomicI64,
}

impl TopicTrafficMeter {
    #[must_use]
    pub fn new() -> Self {
        Self {
            counters: DashMap::new(),
        }
    }

    /// Ensure an accumulator exists for `topic`. Idempotent. After this,
    /// the topic emits zero on every scrape until traffic arrives —
    /// eliminating the gh #115 "no data" gap on idle topics.
    pub fn touch(&self, topic: &str) {
        if topic.is_empty() {
            return;
        }
        self.counters
            .entry(topic.to_string())
            .or_insert_with(|| Arc::new(TopicCounters::default()));
    }

    /// Drop the accumulator. Called on `KafkaTopic` CR delete so orphan
    /// timeseries don't linger on the dashboard.
    pub fn forget(&self, topic: &str) {
        self.counters.remove(topic);
    }

    /// Hot-path from the Produce handler. Auto-touches the topic.
    pub fn record_produce(&self, topic: &str, records: i64, bytes: i64) {
        let tc = self.ensure(topic);
        if records > 0 {
            tc.produce_records.fetch_add(records, Ordering::Relaxed);
        }
        if bytes > 0 {
            tc.produce_bytes.fetch_add(bytes, Ordering::Relaxed);
        }
    }

    /// Hot-path from the Fetch handler. Auto-touches the topic.
    /// `bytes` may be 0 (empty Fetch response) — the accumulator still
    /// gets touched so idle topics keep emitting.
    pub fn record_fetch(&self, topic: &str, records: i64, bytes: i64) {
        let tc = self.ensure(topic);
        if records > 0 {
            tc.fetch_records.fetch_add(records, Ordering::Relaxed);
        }
        if bytes > 0 {
            tc.fetch_bytes.fetch_add(bytes, Ordering::Relaxed);
        }
    }

    fn ensure(&self, topic: &str) -> Arc<TopicCounters> {
        if topic.is_empty() {
            // Return a throwaway accumulator so hot-path callers don't
            // need to nil-check. Cheap: 4 AtomicI64.
            return Arc::new(TopicCounters::default());
        }
        // Fast path: read-only entry access via DashMap.
        if let Some(existing) = self.counters.get(topic) {
            return Arc::clone(&existing);
        }
        Arc::clone(
            &self
                .counters
                .entry(topic.to_string())
                .or_insert_with(|| Arc::new(TopicCounters::default())),
        )
    }

    fn snapshot(&self) -> Vec<TopicSnapshot> {
        self.counters
            .iter()
            .map(|entry| TopicSnapshot {
                topic: entry.key().clone(),
                produce_records: entry.value().produce_records.load(Ordering::Relaxed),
                produce_bytes: entry.value().produce_bytes.load(Ordering::Relaxed),
                fetch_records: entry.value().fetch_records.load(Ordering::Relaxed),
                fetch_bytes: entry.value().fetch_bytes.load(Ordering::Relaxed),
            })
            .collect()
    }
}

#[derive(Debug, Clone)]
struct TopicSnapshot {
    topic: String,
    produce_records: i64,
    produce_bytes: i64,
    fetch_records: i64,
    fetch_bytes: i64,
}

/// Register the four always-emit observable counters against `meter`,
/// each with a callback that walks `traffic`'s snapshot.
///
/// Kept as a free function (rather than an `impl` method) so the
/// callbacks capture an `Arc<TopicTrafficMeter>` cleanly — `&self`
/// methods would fight with the callback's `'static` lifetime bound.
pub fn register_topic_traffic_instruments(meter: &Meter, traffic: Arc<TopicTrafficMeter>) {
    let t = Arc::clone(&traffic);
    meter
        .u64_observable_counter("kaas.produce.records")
        .with_description(
            "Cumulative records produced per topic. Emits at every scrape interval (including 0 for idle topics) so dashboard rate() panels never gap.",
        )
        .with_unit("{record}")
        .with_callback(move |observer| {
            for snap in t.snapshot() {
                observer.observe(
                    to_u64(snap.produce_records),
                    &[KeyValue::new("topic", snap.topic)],
                );
            }
        })
        .build();

    let t = Arc::clone(&traffic);
    meter
        .u64_observable_counter("kaas.produce.bytes")
        .with_description("Cumulative bytes produced per topic. Idle-emit invariant.")
        .with_unit("By")
        .with_callback(move |observer| {
            for snap in t.snapshot() {
                observer.observe(
                    to_u64(snap.produce_bytes),
                    &[KeyValue::new("topic", snap.topic)],
                );
            }
        })
        .build();

    let t = Arc::clone(&traffic);
    meter
        .u64_observable_counter("kaas.fetch.records")
        .with_description("Cumulative records fetched per topic. Idle-emit invariant.")
        .with_unit("{record}")
        .with_callback(move |observer| {
            for snap in t.snapshot() {
                observer.observe(
                    to_u64(snap.fetch_records),
                    &[KeyValue::new("topic", snap.topic)],
                );
            }
        })
        .build();

    let t = Arc::clone(&traffic);
    meter
        .u64_observable_counter("kaas.fetch.bytes")
        .with_description("Cumulative bytes fetched per topic. Idle-emit invariant.")
        .with_unit("By")
        .with_callback(move |observer| {
            for snap in t.snapshot() {
                observer.observe(
                    to_u64(snap.fetch_bytes),
                    &[KeyValue::new("topic", snap.topic)],
                );
            }
        })
        .build();
}

fn to_u64(v: i64) -> u64 {
    // Cumulative counters are non-negative; negative values would be a
    // logic bug upstream (underflow). Saturate at 0 so a corrupt
    // accumulator doesn't produce a bogus 9 exabyte value.
    u64::try_from(v).unwrap_or(0)
}

#[cfg(test)]
#[allow(clippy::unwrap_used, clippy::expect_used)]
mod tests {
    use super::*;

    #[test]
    fn touch_forget_are_idempotent() {
        let m = TopicTrafficMeter::new();
        m.touch("foo");
        m.touch("foo");
        assert!(m.counters.contains_key("foo"));
        m.forget("foo");
        m.forget("foo");
        assert!(!m.counters.contains_key("foo"));
    }

    #[test]
    fn empty_topic_is_ignored() {
        let m = TopicTrafficMeter::new();
        m.record_produce("", 1, 100);
        m.touch("");
        assert!(m.counters.is_empty());
    }

    #[test]
    fn record_produce_and_fetch_accumulate() {
        let m = TopicTrafficMeter::new();
        m.record_produce("foo", 5, 1000);
        m.record_produce("foo", 3, 500);
        m.record_fetch("foo", 4, 800);

        let snaps = m.snapshot();
        assert_eq!(snaps.len(), 1);
        assert_eq!(snaps[0].produce_records, 8);
        assert_eq!(snaps[0].produce_bytes, 1500);
        assert_eq!(snaps[0].fetch_records, 4);
        assert_eq!(snaps[0].fetch_bytes, 800);
    }

    #[test]
    fn snapshot_returns_all_topics() {
        let m = TopicTrafficMeter::new();
        m.record_produce("foo", 1, 100);
        m.record_produce("bar", 2, 200);
        m.touch("baz"); // idle-emit case

        let snaps = m.snapshot();
        assert_eq!(snaps.len(), 3);
    }

    #[test]
    fn to_u64_saturates_negatives() {
        assert_eq!(to_u64(-1), 0);
        assert_eq!(to_u64(0), 0);
        assert_eq!(to_u64(42), 42);
    }
}
